matplotcheck/raster.py

```diff
@@ -56,112 +56,174 @@ def assert_colorbar_range(self, crange):
+        #  better if it returns a dictionary with the key being each
+        #  label being tested and the value being a boolean (True if there
+        #  is a match, False if there is no match)
+
+        labels_dict = self._check_label(labels, expected_labels)
+
+        # labels_check = [
+        #     self._check_label(label, expected_labels[i])
+        #     for i, label in enumerate(labels)
+        # ]
+
+        # TODO: this now becomes a bit more tricky but we need to 1) test that
+        # each label is true - if one is false, then return useful message
+        # stating which label is incorrect.
+
+        # Pull out any labels that failed the above test for final printing
+        # below
+        bad_labels = {
+            key: labels_dict[key]
+            for key in labels_dict
+            if not labels_dict[key]
+        }
+
+        # TODO: raise assertion error (value error?) and print out a list of
+        #  labels that are wrong ONLY if some are wrong
+        if bad_labels:
+            # get just the labels that are
+            bad_keys = [str(a_key) for a_key in bad_labels.keys()]
+            raise ValueError(
+                "Oops. It looks like atleast one of your legend "
+                "labels is incorrect. Double check the "
+                "following label(s): {"
+                "}".format(bad_keys)
+            )
+
+        # Check that each legend entry label is in one of expected_labels
+        # assert all(
+        #     labels_check
+        # ), "Provided legend labels don't match labels found."
+
+        # TODO: once we get the above working, let's then add another layer
+        #  where we grab the RGB values and also add that to the dictionary
+        # At that point we can test whether the colors in the plot array, map
+        # to the legend patch colors and in turn the expected image
+
+        # Check that expected and actual arrays data match up
+        assert np.array_equal(
+            im_data, im_expected
+        ), "Expected image data doesn't match data in image."
+
+        # TODO: warning -- proj_create: init=epsg:/init=IGNF: syntax not
+        #  supported in non-PROJ4 emulation mode - where is this coming from?
+
+        # IMAGE TESTS/HELPER FUNCTIONS
+
+    def get_legend_labels(self):
+        """Return labels from legend in a list
+
+        Returns
+        -------
+        labels: List
+            List of labels found in the legend of a raster plot.
+        """
 
         # Retrieve legend
         legends = self.get_legends()
         assert legends, "No legend displayed"
 
-        # Retrieve legend entries and find which element of all_label_options
-        # matches that entry
-        legend_dict = {}
-        for p in [
-            p
-            for sublist in [leg.get_patches() for leg in legends]
-            for p in sublist
-        ]:
-            label = p.get_label().lower()
-            legend_dict[p.get_facecolor()] = self._which_label(
```

Comment:
i thijnk we can use this code to get both the label and the associated label color and add both to the dictionary that i created in check_label. this might require a slight restructure however but i thought we could implement this iteratively.